### PR TITLE
[Concept Entry] Created Terms folder and added Euclidean algorithm 

### LINKED ISSUE #2454: [Euclidean Algo] Adding terms folder inside the algorithm folder
### Reason/inspiration (optional)

- Adding terms folder inside the algorithm folder
- Adding the first entry inside the terms folder i.e. Euclidean algorithm

### Entry Type (select all that apply)

New Topic Entry

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [X] Issue is assigned

### PR DESCRIPTION
**Pull request by dakshdeepHERE**
### Type of Change

- Adding a new entry
- Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

Fixes #2454

### CODE REVIEW

**Comment by dakshdeepHERE** [comment]
@SSwiniarski No changes required from your side ?

**Comment by SSwiniarski** [comment]
@dakshdeepHERE Have patience, it hasn't been reviewed yet.

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (CBID2) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:50]: [Nitpick] According to the docs style guide, pronouns can't be used. 
```suggestion
In this method, the aim is to compare both 'a' and 'b'. Whichever number is greater,  the smaller number is subtracted from the larger number. It also updates the value of the larger number accordingly. If 'a' is greater than 'b', replace 'a' with 'a - b', and if 'b' is greater than 'a', replace 'b' with 'b - a'. Repeat this step until 'a' becomes equal to 'b'. Once 'a' and 'b' are equal, return 'a' as the answer. At this point, the value of 'a' (which is equal to 'b') represents the highest common factor (HCF) of the original values of 'a' and 'b'.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (CBID2) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:13]: [Nitpick] See my feedback on [this line](https://github.com/Codecademy/docs/pull/2455#discussion_r1226023651). They both share the same problem. 
```suggestion
The `Euclidean algorithm` is a recursive algorithm that allows us to find the highest common factor (HCF) of two numbers. It is based on the observation that if we have two numbers, say A and B, and if B divides A evenly (leaving no remainder), then B is the HCF of A and B. In other words, the larger number is divisible by the smaller number without any remainder.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (CBID2) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:95]: [Nitpick] See my feedback on [this line](https://github.com/Codecademy/docs/pull/2455#discussion_r1226023651). They share the same problem. 
```suggestion
If we debug the code above step by steps then:

- Since 'b' is greater than 'a' (20 > 15), we replace 'b' with 'b - a', which gives us 'b = 20 - 15 = 5'.
- Now 'a' is still 15, and 'b' is 5.
- We proceed to replace 'a' with 'a - b', resulting in 'a = 15 - 5 = 10'.
- The values are now 'a = 10' and 'b = 5'.
- We continue the process and replace 'a' with 'a - b', giving us 'a = 10 - 5 = 5'.
- Now 'a' and 'b' are both equal to 5.
- At this point, the while loop exits, and we return 'a' as our answer, which is 5.
- Therefore, the highest common factor (HCF) of (15, 20) is 5.

## Method 3

In this method we will use [recursion](https://www.codecademy.com/learn/discrete-math-recurrence-relations). The method takes two integer parameters 'a' and 'b' . Then returns an integer as the result. The gcd method uses a `recursive` approach to implement the `Euclidean algorithm`. It first checks if `b` is equal to 0. If it is, then it means that `a` is the GCD, and it returns `a` as the result. If `b` is not 0, it means there is a remainder when `a` is divided by `b`. In this case, the method calls itself recursively with the arguments `b` and `a % b`. This recursive call continues until `b` becomes 0, at which point the base case is triggered, and the GCD is found.
```

**Review by CBID2 — CHANGES_REQUESTED** [review]
Hi @dakshdeepHERE! :) I gave some feedback on your entry. I hope you find them helpful (note: I also recommend running your entry through Grammarly or spellcheck for errors! 😊🙏🏽

**Comment by dakshdeepHERE** [comment]
Alright @SSwiniarski and @CBID2  I mostly try reverify everything through my grammarly extension already. Thanks for the advise is there anything else I can improve? ❤️🙏🏻

**Comment by CBID2** [comment]
> Alright @SSwiniarski and @CBID2 I mostly try reverify everything through my grammarly extension already. Thanks for the advise is there anything else I can improve? ❤️🙏🏻

Outside of updating your branch, nothing on my end @dakshdeepHERE. :)

**Comment by dakshdeepHERE** [comment]
@CBID2 Check the changes I removed all the pronouns and refactored the sentences too with appropriate grammar. Let me know if further changes are required.

**Comment by dakshdeepHERE** [comment]
Also, I need suggestions on the last part. In the future, I will continue following this trend with all my algorithm contributions. It is important to understand the differences between using different approaches and how they affect the respective time complexity.

**Review by CBID2 — CHANGES_REQUESTED** [review]
The third part looks ok @dakshdeepHERE. I just gave a suggestion on how to improve the final paragraph.

**Comment by dakshdeepHERE** [comment]
Done @CBID2

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (CBID2) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:13]:
```suggestion
The `Euclidean algorithm` is a recursive algorithm that allows developers to find the highest common factor (HCF) of two numbers. It is based on the observation that if two numbers are given, say A and B, where B divides A evenly (leaving no remainder), then B is the HCF of A and B. In other words, the larger number is divisible by the smaller number without any remainder.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (CBID2) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:19]:
```suggestion
In the basic approach, the purpose is to find the GCD. To do this, find the minimum value between the two given numbers. Then, divide it by both numbers. If the number does not divide both numbers evenly, decrease it by `1` and continue dividing. Lastly, repeat this process until the minimum value can divide both numbers evenly. At this point, the minimum value is the HCF (highest common factor).
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (CBID2) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:97]:
```suggestion
For more information on recursion, refer to this [resource](https://www.codecademy.com/learn/java-algorithms/modules/recursion-apcs/cheatsheet).
```

**Review by CBID2 — CHANGES_REQUESTED** [review]
Just found some more things to improve @dakshdeepHERE. Keep it up! :)

**Review by CBID2 — APPROVED** [review]
Looks good to me

**Comment by dakshdeepHERE** [comment]
Thanks for the suggestions @CBID2 My next submissions won't have these pronouns.

**Comment by dakshdeepHERE** [comment]
Sorry for the tag, Can you look into this @SSwiniarski @caupolicandiaz its been like a week.

**Comment by SSwiniarski** [comment]
@dakshdeepHERE I'm looking into this now.

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:3]:
```suggestion
Description: 'A simple and efficient method for finding the highest common factor (HCF), also known as the greatest common divisor (GCD), of two numbers.'
```
This because of the way terms are displayed on the parent page. They show up as [title] - [description] so having the title of the entry in the description will cause a redundancy.

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:8]:
```suggestion
  - 'Arithmetic'
```
Reminder, tags need to come from [this list](https://github.com/Codecademy/docs/blob/main/documentation/tags.md).

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:133]:
```suggestion
  - The first code uses a simple iterative approach to find the greatest common divisor (GCD) of two numbers. It starts by finding the minimum value between `a` and `b`, then iterates from that value down to 1, checking if it divides both `a` and `b`. Therefore, the time complexity of this code is **O(min(a, b))**.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:137]:
```suggestion
  - The second code also calculates the GCD using an iterative approach known as the Euclidean algorithm. It repeatedly subtracts the smaller number from the larger number until the two numbers become equal (the GCD). The time complexity of this algorithm depends on the number of iterations required to reach the GCD. In the worst case, where one number is a multiple of the other, the time complexity is **O(max(a, b))**.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:141]:
```suggestion
  - The third code is an optimized version of the Euclidean algorithm that uses recursion. It calculates the GCD by repeatedly taking the modulus of `a` with `b` and calling itself with the new values (`b` and `a%b`) until `b` becomes 0. The time complexity of this optimized Euclidean algorithm is **O(log(max(a, b)))** since the algorithm reduces the values quickly by taking the modulus.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:145]:
```suggestion
- Code 1: **O(min(a, b))**
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:146]:
```suggestion
- Code 2: **O(max(a, b))**
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:147]:
```suggestion
- Code 3: **O(log(max(a, b)))**
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:149]:
```suggestion
> **Note:** These time complexities typically represent the worst-case scenarios and assume that the `a` and `b` values are relatively large. In practice, the actual time taken by the algorithms can vary depending on the input values.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:21]: Can we do these examples in pseudocode rather than Java?

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:52]: Can we do this in pseudocode rather than Java?

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (SSwiniarski) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:99]: Can we do this in pseudocode rather than Java?

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@dakshdeepHERE I did a review and left a few comments. I also have a big overall comment: Because general entries are supposed to be language agnostic, can we write the examples in pseudocode rather than a specific programming language so this entry doesn't assume a knowledge of Java.

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (dakshdeepHERE) [reply content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:3]: alright

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (dakshdeepHERE) [reply content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:8]: Got it. We have to open a separate PR to add tags in tag.md?

**Comment by dakshdeepHERE** [comment]
@SSwiniarski Check the changes once to confirm if there's more changes needed. Also I will keep this check that I don't use tag **java** and instead of it use **pseudo** for examples. I thought the whole algorithm is coded in java so using **java** tag would be useful.

**Comment by dakshdeepHERE** [comment]
@SSwiniarski I've seen the main algorithm.md and there the implementation of algorithm was done in python that's I did it in java.
image

**Comment by SSwiniarski** [comment]
I'll defer to the other maintainers: @yangc95, @KTom101, @caupolicandiaz: do you think it's all right for this general example to be written in Java?

**Comment by dakshdeepHERE** [comment]
Sure @SSwiniarski because I was preparing the algorithms in Java for future entries. If we device anything else then I will follow that.

**Comment by KTom101** [comment]
@SSwiniarski @dakshdeepHERE I feel it's alright as long as the language is specified. A couple of our [Search Algorithms](https://www.codecademy.com/resources/docs/ai/search-algorithms) (for example [BFS](https://www.codecademy.com/resources/docs/ai/search-algorithms/breadth-first-search)) do the same thing.

**Comment by dakshdeepHERE** [comment]
@KTom101 yes I was using java tags in the code. Should I also specify at beginning about the language we are going to use in the algorithm entries or the tag for each code is enough?

**Comment by KTom101** [comment]
@dakshdeepHERE You can write it before the example similar to how we do in other entries. 
You can write "The following shows how to implement ... in Java" filling in the "..." with what you're showing.

**Comment by dakshdeepHERE** [comment]
> You can write "The following shows how to implement ... in Java"

Understood I will do this before examples or mention it in the explanations down below it.

**Comment by dakshdeepHERE** [comment]
@SSwiniarski so should we continue with java?

**Comment by SSwiniarski** [comment]
@dakshdeepHERE Let's go ahead with Java, with the additions @KTom101 suggested.

**Comment by dakshdeepHERE** [comment]
Alright @SSwiniarski should I change the tags to java now?

**Comment by dakshdeepHERE** [comment]
@SSwiniarski Done. Is there any other changes you want me to do?

**Comment by SSwiniarski** [comment]
@dakshdeepHERE Add @KTom101's suggestion by adding sentences like "The following code illustrates this method in Java." before your code blocks.

**Comment by dakshdeepHERE** [comment]
@SSwiniarski alright

**Comment by dakshdeepHERE** [comment]
@SSwiniarski Can you check the changes now.

**Review by SSwiniarski — APPROVED** [review]
@dakshdeepHERE This looks good for a second review.

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (caupolicandiaz) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:127]:
```suggestion
- In this recursive cycle, when the code enters the `if` statement, the condition `b == 0` is satisfied. Therefore, the function will return `a`,  which has a value of `50`, the highest common factor for the original arguments.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (caupolicandiaz) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:129]:
```suggestion
## Time Complexities
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (caupolicandiaz) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:133]:
```suggestion
  - The first example uses a simple iterative approach to find the greatest common divisor (GCD) of two numbers. It starts by finding the minimum value between `a` and `b`, then iterates from that value down to 1, checking if it divides both `a` and `b`. Therefore, the time complexity of this code is *O(min(a, b))*.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (caupolicandiaz) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:137]:
```suggestion
  - The second example also calculates the GCD using an iterative approach known as the Euclidean algorithm. It repeatedly subtracts the smaller number from the larger number until the two numbers become equal (the GCD). The time complexity of this algorithm depends on the number of iterations required to reach the GCD. In the worst case, where one number is a multiple of the other, the time complexity is *O(max(a, b))*.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (caupolicandiaz) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:141]:
```suggestion
  - The third example is an optimized version of the Euclidean algorithm that uses recursion. It calculates the GCD by repeatedly taking the modulus of `a` with `b` and calling itself with the new values (`b` and `a%b`) until `b` equals `0`. The time complexity of this optimized Euclidean algorithm is *O(log(max(a, b)))* since the algorithm reduces the values quickly by taking the modulus.
```

File: content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md
Comment (caupolicandiaz) [inline content/general/concepts/algorithm/terms/euclidean-algorithm/euclidean-algorithm.md:147]:
```suggestion
- Method 1: *O(min(a, b))*
- Method 2: *O(max(a, b))*
- Method 3: *O(log(max(a, b)))*
```

**Review by caupolicandiaz — CHANGES_REQUESTED** [review]
@dakshdeepHERE I've completed my initial review. Please update the entry, and tag me when you've made the changes. Thanks

**Comment by dakshdeepHERE** [comment]
@caupolicandiaz All the changes are done. Can you check once if there's any other changes needs to be done?

**Review by caupolicandiaz — APPROVED** [review]
@dakshdeepHERE @SSwiniarski I pushed a couple of additional small edits. I think this is good to go.

**Comment by SSwiniarski** [comment]
@dakshdeepHERE Congrats on the new entry. Live link here: https://www.codecademy.com/resources/docs/general/algorithm/euclidean-algorithm

image

**Pull request merged**
